### PR TITLE
Fix data_source not updated on re-import and refactor `reimport_metadata`

### LINKED ISSUE #4726: Re-imports don't update `data_source` flex attr
### Problem

When importing an album or a singleton that was previously imported already, the `data_source` flexible attribute is not updated. It stays as **initially imported**.

### Setup

* OS: ---
* Python version: ---
* beets version: 1.6.1 dev-latest
* Turning off plugins made problem go away (yes/no): no

### Proposed solution 

`data_source` needs to be excluded in `reimport_metadata()`

### PR DESCRIPTION
**Pull request by JOJ0**
## Description

- Fixes #4726 
- Refactors preserving of flexible attributes in the importer's `reimport_metadata()` method.
- New logging message when flexible attributes are excluded from the preserving process and _actually_ set to a new value.
- Reformats all existing logging messages and streamline their looks.
- Adds a new unittest that checks if the data_source album field is _not_ reset to its original value on a reimport.

## To Do

- [X] ~Documentation.~
- [X] Changelog.
- [X] Tests.

### CODE REVIEW

**Comment by JOJ0** [comment]
Hi @sampsyo, I would love to get a review from you. This might be tedious to look at in the final diff. I suggest looking at it commit-wise. The actual technical change you'll find in the first one: https://github.com/beetbox/beets/pull/4735/commits/bf29e4fb65c178ce88d622bed630da01363f9321

Thanks!

If you want I can post some logging examples comparing "before" and "after". If that might help?

File: beets/importer.py
Comment (sampsyo) [inline beets/importer.py:818]: Let's make these into module-level constants. May I suggest names like `REIMPORT_FRESH_FIELDS_ITEM` or something to make it clear where they will be used?

File: beets/importer.py
Comment (sampsyo) [inline beets/importer.py:835]: I was thinking a little bit about how to simplify this stuff, which gets surprisingly intricate for what it does. I sketched up this possibly-broken list comprehension idea; do you think it's anywhere close to what we want?

```py
old_fields = dict(replaced_album._values_flex)
overwritten_fields = [k for k in old_fields
                      if k in REIMPORT_FRESH_FIELDS_ITEM
                      and old_fields.get(k) != self.album.get(k)]
if overwritten_fields:
    log.debug("Not preserving attributes: {}", overwritten_fields)
    for key in overwritten_fields:
        del old_fields[key]
```

File: beets/importer.py
Comment (sampsyo) [inline beets/importer.py:874]: Since we end up doing more or less the same thing for items and albums, it would be really cool to reduce some of the code duplication that results. Maybe a very small helper function could be the right thing here?

**Review by sampsyo — COMMENTED** [review]
Looking great so far!! Thank you for tackling this! I've left a couple of low-level code comments.

**Comment by JOJ0** [comment]
> Looking great so far!! Thank you for tackling this! I've left a couple of low-level code comments.

Very much appreciated! That was exactly the review I needed! Awesome. See my changes and comments within.

File: beets/importer.py
Comment (JOJ0) [reply beets/importer.py]: Awesome! Thanks for this great suggestion. Have a look at https://github.com/beetbox/beets/pull/4735/commits/8bac74480c6d0ffd57175808f0e52f187ce9342e

As often in IT, the hardest part was "naming things". I'm open for suggestions if you think some of my var or docstring names sound clunky.

Comment (JOJ0) [reply beets/importer.py]: It's funny that you used the term "preserving" - I'm having the exact same thought since I started to refactor this thing. I kind of found the term "reimported" in the log messages kind of confusing.

What do you think about to change all existing log messages to use that term instead of the currently used "Reimported...."?

Comment (JOJ0) [reply beets/importer.py]: A comparision of possible ways to log this:

1) Original version plus logging not-preserved values:

```
Setting album flexible attribute data_source: match_source for 1, /tmp/tmpu5nwti8q/libdir/singletons
Sending event: database_change
Sending event: database_change
Reimported album: added 4242.0, flexible attributes ['foo'] from album 1 for /tmp/tmpu5nwti8q/libdir/singletons
Reimported item added 4747.0 from item 1 for /tmp/tmpu5nwti8q/libdir/singletons/tïtle 0.mp3
Reimported item flexible attributes ['baz'] from item 1 for /tmp/tmpu5nwti8q/libdir/singletons/tïtle 0.mp3
```

2) Keeping the term "Reimported":

```
Not reimported album 1 flexible attribute: data_source. New value: match_source, Path: /tmp/tmpddrw2quc/libdir/singletons
Sending event: database_change
Sending event: database_change
Reimported album 1 attribute: ['added'] from album 1, Path: /tmp/tmpddrw2quc/libdir/singletons
Reimported album 1 flexible attributes: ['foo'] from album 1, Path:/tmp/tmpddrw2quc/libdir/singletons
Reimported item 1 attribute: ['added'] from item 1, Path: /tmp/tmpddrw2quc/libdir/singletons/tïtle 0.mp3
Reimported item 1 flexible attributes: ['baz'] from item 1, Path: /tmp/tmpddrw2quc/libdir/singletons/tïtle 0.mp3
```

3) Use the term "Preserved/Preserving"

```
Not preserving album 1 flexible attributes: ['data_source']. Path: /tmp/tmpda4c_4c3/libdir/singletons
Sending event: database_change
Sending event: database_change
Preserved album 1 attribute: ['added'], Path: /tmp/tmpda4c_4c3/libdir/singletons
Preserved album 1 flexible attributes: ['foo'], Path:/tmp/tmpda4c_4c3/libdir/singletons
Preserved item 1 attribute: ['added'], Path: /tmp/tmpda4c_4c3/libdir/singletons/tïtle 0.mp3
Preserved item 1 flexible attributes: ['baz'], Path: /tmp/tmpda4c_4c3/libdir/singletons/tïtle 0.mp3
```

I removed the "from album/item <ID>" in that last proposal. I am assuming that never ever on a reimport the ID of the reimported albums/items will change. Do you have any doubts in that assumption?

Comment (sampsyo) [reply beets/importer.py]: Yeah, 100% agree that the current log messages are not well thought-out. They are mainly there because they are better than not logging anything at all.

FWIW, I think it would be nice to include the word "reimport" _somewhere_ in the messages, so people know that this basic behavior has been invoked:
https://beets.readthedocs.io/en/stable/reference/cli.html#reimporting

But yeah, "preserving" is much better wording for what we're doing to the metadata during a reimport.

I think you're right that the ID will never change on a reimport! I am 60% confident in that assertion. :smiley:

**Review by sampsyo — APPROVED** [review]
Awesome!!! Seems pretty legit to me. If this does what we want, I think it's basically good to go (modulo bikeshedding the log messages).

File: beets/importer.py
Comment (JOJ0) [reply beets/importer.py]: Together with my own 60%, we'd  be at 120%, that's more than enough! :laughing:

Comment (JOJ0) [reply beets/importer.py]: This is my final version of the log messages.

```
Reimported album 1. Not preserving flexible attributes ['data_source']. Path: /tmp/tmpoo53zw7q/libdir/singletons
Sending event: database_change
Sending event: database_change
Reimported album 1. Preserving attribute ['added']. Path: /tmp/tmpoo53zw7q/libdir/singletons
Reimported album 1. Preserving flexible attributes ['foo']. Path: /tmp/tmpoo53zw7q/libdir/singletons
Reimported item 1. Preserving attribute ['added']. Path: /tmp/tmpoo53zw7q/libdir/singletons/tïtle 0.mp3
Reimported item 1. Preserving flexible attributes ['baz']. Path: /tmp/tmpoo53zw7q/libdir/singletons/tïtle 0.mp3
```

**Comment by JOJ0** [comment]
Thanks @sampsyo for the input. Log messages have been bikeshed! :-)) Merging.

File: beets/importer.py
Comment (JOJ0) [reply beets/importer.py]: Especially with real-world data it makes sense to start with capital letters to make things "digestable" in this "information salad":

```
Reimported album 29. Not preserving flexible attributes ['data_source']. Path: /home/jojo/Documents/Sync/beets_test_music/Uncategorized/A - D/Chimpo - 2022 -  On The Dial
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Reimported album 29. Preserving attribute ['added']. Path: /home/jojo/Documents/Sync/beets_test_music/Uncategorized/A - D/Chimpo - 2022 -  On The Dial
Reimported album 29. Preserving flexible attributes ['toplevel']. Path: /home/jojo/Documents/Sync/beets_test_music/Uncategorized/A - D/Chimpo - 2022 -  On The Dial
Reimported item 137. Preserving attribute ['added']. Path: /home/jojo/Documents/Sync/beets_test_music/Uncategorized/A - D/Chimpo - 2022 -  On The Dial/01 - Chimpo - On The Dial - Rig Doctor (2022).flac
Reimported album 137. Not preserving flexible attributes ['data_source']. Path: /home/jojo/Documents/Sync/beets_test_music/Uncategorized/A - D/Chimpo - 2022 -  On The Dial/01 - Chimpo - On The Dial - Rig Doctor (2022).flac
```

Comment (JOJ0) [reply beets/importer.py]: Whoops, thanks to that log msg post above, I see  that there is a tiny bug. Last line should rather say "Reimported item 137...." Fixed.

**Comment by sampsyo** [comment]
Excellent!

**Pull request merged**
